src/main/flight/pid.c

```diff
@@ -487,6 +486,13 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_
 #ifdef USE_GPS_RESCUE
     angleTarget += gpsRescueAngle[axis] / 100.0f; // Angle is in centidegrees, stepped on roll at 10Hz but not on pitch
 #endif
+#ifdef USE_POS_HOLD_MODE
+    if (FLIGHT_MODE(POS_HOLD_MODE)) {
+        angleTarget += autopilotAngle[axis]; // autopilotAngle in degrees
```

Comment:
THis mixing of currentPidSetpoint and 'magic' angle variables is confusing. Isn;t it possible to feel desired angle into pidLevel ?

Reply:
Good point.
While still in the Position Hold flight mode, as soon as the pilot moves sticks outside the deadband range, the previously active Position Hold `autopilotAngle[axis]` angle targets are no longer relevant, and are replaced with 'normal' pilot-generated Angle Mode `angleTarget` values.

At this moment, despite being in the Position Hold flight mode, the pilot is flying in Angle mode.  The autopilot code sends zero for `angleTarget` when the sticks get moved outside the deadband.

Conversely, when the sticks go inside the deadband range, pid.c receives a zero setpoint, due to the deadband in rc.c, leading to a zero AngleMode `angleTarget`.  At the same time, the autopilot code sends an alternative  `angleTarget` via `autopilotAngle[axis]`.

When landing with small stick corrections, there may be many transitions of this kind.

We want smooth transitions from Pilot control to and from Position Hold.  

The Position Hold `angleTarget` value has some lowpass smoothing tail-off from the lowpass filters on D.  I think also that there is some residual tail-off in the normal `angleTarget` from the pilot as well.

By adding the two, and zeroing them appropriately, we get a smooth transition between the two behaviours.

An alternative approach could be to not have a deadband in rc.c at all, and just over-write `angleTarget` pid.c when the sticks are inside the deadband.  This may not be as smooth in the transitions, but probably would simplify the code.

The reason I did it the current way was that I just copied the method from gps_rescue.  Not because I thought it was the best method, but because I knew it worked :-) 

I'm open to better or simpler solutions.

Reply:
For now, it would be enough to separate angle input mixing (RC, rescue, poshold) into separate function. Then it would be more clear what actual angle setpoint is; it can be improved in future.
Now the data flow is mostly hidden, you need to review multiple files to see where is the value coming from.

> We want smooth transitions from Pilot control to and from Position Hold.

Only upsampling filter is applied, 


> At this moment, despite being in the Position Hold flight mode, the pilot is flying in Angle mode. The autopilot code sends zero for angleTarget when the sticks get moved outside the deadband.

> Conversely, when the sticks go inside the deadband range, pid.c receives a zero setpoint, due to the deadband in rc.c, leading to a zero AngleMode angleTarget. At the same time, the autopilot code sends an alternative angleTarget via autopilotAngle[axis].

This logic is spread over three files, mixed with lot of other RC magic

> An alternative approach could be to not have a deadband in rc.c at all, and just over-write angleTarget pid.c when the sticks are inside the deadband. This may not be as smooth in the transitions, but probably would simplify the code.

This will probably make things even worse. Angle won't start from zero (deadband must match) and data flow will be even more magic.
But poshold (poshold-related function) can take RC as input, conditionally modify it and pass resulting setpoint down - to angle controller.

> The reason I did it the current way was that I just copied the method from gps_rescue. Not because I thought it was the best method, but because I knew it worked :-)

It was quite acceptable for rescue - it is exceptional situation and overriding normal dataflow isn't that bad in such condition.

> I'm open to better or simpler solutions.

Maybe there in no simple solution. But adding layer over layer of hacks will quickly lead to unusable codebase. At least keep them as isolated as possible.